### PR TITLE
Direct export to excel format

### PR DESCRIPTION
**Pull request by stefano**
Implements https://github.com/metabase/metabase/issues/4417

Adds docjure (https://github.com/mjul/docjure) as a dependency to write excel files.

As part of the PR I refactored CSV/JSON exports to reduce code duplication and make it easier to add more export formats in the future.

Question: is the API documentation automatically generated or should I update it manually?

### CODE REVIEW

File: project.clj
Comment (tlrobinson) [inline project.clj]: Could you put this between `crypto-random` and `environ` so it's alphabetical?

File: src/metabase/api/card.clj
Comment (tlrobinson) [inline src/metabase/api/card.clj]: We could probably shorten `export-format-name` to `export-format`.

**Review by tlrobinson — COMMENTED** [review]
Cool thanks for doing this! I'll let @camsaul review the backend more thoroughly but it's looking good.

File: src/metabase/api/card.clj
Comment (camsaul) [inline src/metabase/api/card.clj]: I would also add a schema here to validate `export-format`

```clojure
{parameters    (s/maybe su/JSONString)
 export-format (s/enum "csv" "json" "xlsx")}
```

File: src/metabase/api/dataset.clj
Comment (camsaul) [inline src/metabase/api/dataset.clj]: Just do `defn-` instead of `defn ^:private`

File: src/metabase/api/embed.clj
Comment (camsaul) [inline src/metabase/api/embed.clj]: same thing, probably just shorten this to `export-format` and add a schema validation here

File: src/metabase/api/public.clj
Comment (camsaul) [inline src/metabase/api/public.clj]: same thing, probably just shorten this to `export-format` and add a schema validation here

File: src/metabase/models/query_execution.clj
Comment (camsaul) [inline src/metabase/models/query_execution.clj]: @salsakran do we want to stop differentiating between whether a query is downloaded as CSV/JSON/Excel? Seems like it might be useful for usage stats

File: test/metabase/api/card_test.clj
Comment (camsaul) [inline test/metabase/api/card_test.clj]: Just use `count` to get the length of something

```clojure
#(> (count %) 0)
```

**Comment by camsaul** [comment]
I'm not convinced that a test just to make sure the endpoint returns a string with at least one byte is really a good enough test. Can `docjure` be used to parse the file that comes back and transform it to some sort of raw (perhaps CSV) output that we can verify?

**Review by camsaul — CHANGES_REQUESTED** [review]
Code cleanup and better tests

File: src/metabase/models/query_execution.clj
Comment (salsakran) [inline src/metabase/models/query_execution.clj]: strong 👎 
we should add an excel download option.

**Comment by stefano** [comment]
> Can docjure be used to parse the file that comes back and transform it to some sort of raw (perhaps CSV) output that we can verify?

I can use docjure to parse back the excel file and check the cell contents, but the tests http client returns the binary excel file as a string, which corrupts the file (because it converts it to utf8 and bytes that can't be converted get corrupted). I need to see if the http client has an option to return a byte stream or a byte array.

**Comment by stefano** [comment]
@camsaul I've addressed all the code review issues.

**Comment by camsaul** [comment]
Looks good @stefano. There were a couple of linter issues I needed to fix so I opened PR #4922 based on this one so I can make the last few changes. Once CI passes I'll merge that one in.

**Pull request merged**
